### ISSUE TITLE
should may be datalad run record not include empty fields?

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#2985]
ATM run record in commits message would  include
```
     "chain": [],
     "inputs": [],
     "outputs": [],
```
which add no information and only make `git log` longer etc.  So I wondered if may be record could contain only non-default/empty/degenerate fields, e.g. instead of

```json
    {
     "chain": [],
     "cmd": "git-sedi 'assert_true(\\( *.* == .*\\) *)$' 'assert \\1'",
     "exit": 0,
     "inputs": [],
     "outputs": [],
     "pwd": "."
    }
```
become "more to the point"

```json
    {
     "cmd": "git-sedi 'assert_true(\\( *.* == .*\\) *)$' 'assert \\1'",
    }
```
where expected defaults (running from root of dataset and exit with 0 status) were suppressed.

I think this would help to promote "datalad run" for general practice, whenever now a bit length and largely "not interesting" record might not look as appealing.  What do you think?

**Comment by yarikoptic** [comment]
barely related: BTW, note that here I used `git-sedi` which is a shell function within my `.bashrc`... not sure if somehow we could/should warn about using aliases/functions etc which others are likely not to have.

**Comment by bpoldrack** [comment]
Hm. Not sure. At least as far as defaults are concerned I'd prefer being explicit. If default changes or datalad goes away the record would lack potentially crucial information or even distort it otherwise. So, in that case `"exit": 0` and `"pwd": "."` should be included I think.

> if somehow we could/should warn about using aliases/functions etc which others are likely not to have.

I see the point, but I don't think we can do anything about it making a meaningful difference. If you don't include your environment (`datalad-container`), you can always end up with something not reproducible. In this regard an alias is not different from a symlink or just any non-standard software you call. Even with standard software you wouldn't record the version without using a container.
So, to me it's just up to you to decide on the extend of reproducibility by using `run` vs. `containers-run`.

**Comment by yarikoptic** [comment]
> Hm. Not sure. At least as far as defaults are concerned I'd prefer being explicit. If default changes or datalad goes away the record would lack potentially crucial information or even distort it otherwise. So, in that case `"exit": 0` and `"pwd": "."` should be included I think.

As long as there is a clear description of expected fields and their defaults -- we would be as good.  For any backward incompatible changes (e.g. removal of the fields) we should anyways introduce a field stating the version of our "run spec", so we ourselves would know what version of the record we are dealing with and what our assumptions were. non-`"exit": 0` runs should be outliers anyways and cannot happen unless you explicitly call additional `save` (the point is that those should be sparse to run about and thus actually seeing "exit" field being present could help visually to alert to those runs in the history). `"pwd": "."` - what do you see it possibly to change to in the future as another default?

**Comment by bpoldrack** [comment]
> what do you see it possibly to change to in the future as another default?

I don't see any. ;-)
And yes, of course versioning of the spec would help. But the point still stands.
To me it's about the conceptional approach not about those particular fields. I'd prefer a dataset (and its committed information) to be as self-contained as possible. If we can have a record that can be interpreted without datalad or its documentation still being around, then we should go for it. It's one thing to have defaults in the software to determine the behavior of commands. It's another thing to have defaults in the software, that determine the interpretation of stored data. If we are to artificially create holes in the record to be filled by the software instead, it just feels wrong to me. Even a version number will be of no use, if you can't find what that actually means anymore. I'm aware there are limits we can't avoid. And those two examples might not be that critical (although you only know in the future whether that's true and that's kind of the point). My view is just: If there is actual information - include it, not try to conclude it. If there's really no information, I'm fine with leaving the field out altogether.

**Comment by yarikoptic** [comment]
a short comment re self-contained and " If there is actual information - include it, not try to conclude it. " - in general agree.  But ATM with empty "inputs" and "outputs" even if command had inputs and obviously outputs (hence a commit) - it is confusing for uninformed user what those are about since clearly command did have outputs.

**Comment by bpoldrack** [comment]
Agree. Those have no *actual* information in the example record and can be left out. Actually even worse in that case, since it can indeed suggest the wrong thing (no info on what was the output vs. there was no output).

**Comment by kyleam** [comment]
> I think this would help to promote "datalad run" for general practice, whenever now a bit length and largely "not interesting" record might not look as appealing.

I'm not convinced that that would have any real effect on adoption, but conversely I don't imagine that it'd hurt either.  

I'm on board with dropping empty values for `inputs`, `outputs`, and `chain`.  I would weakly vote to keep `pwd` and `exit`.  I worry that `cmd` will get lonely.
